### PR TITLE
August 2024 UI and usability improvements

### PR DESCRIPTION
**Pull request by MatteoBiasi**
This PR includes the the following issues:
- #532 
- #533
- #573
- #526 
- #524 
- #527 
- #525 
- #575

### CODE REVIEW

File: databrowser/src/config/builder/tourism/location.ts
Comment (sseppi) [inline databrowser/src/config/builder/tourism/location.ts:189]: @a-crea did you forget to remove the code you added for testing or shall we remove the comment // FIXME?

**Comment by gappc** [comment]
@MatteoBiasi @a-crea general remark for this PR: it is not up-to-date with the current `development` branch, please bring it up-to-date

**Comment by gappc** [comment]
In the TableView, the new header has lots of white spaces (#575 is about a compacter TableView) and the select-bi / buttons / search input have different heights (small issue). See image below.

image

**Comment by gappc** [comment]
DetailView: referenced attributes are not visible at all, when the `Load referenced attributes` toggle is deactivated, see images below

Referenced attributes visible (correct)
image


Referenced attributes not visible at all (bug?)
image

**Comment by gappc** [comment]
Matter of taste: lots of white space in between Table and Toolbox, see image below

image

**Comment by gappc** [comment]
Proposal for referenced values: clicking on the reference link should open the data in a new tab, see image below.

image

**Comment by gappc** [comment]
DetailView for dataset with additional properties: the accordion arrow icon in the menu on the left is not visible on initial page load, see image below.

Example URL: `BASE_URL/dataset/detail/tourism/v1/ODHActivityPoi/echarging_vek:hyc_150_19bz00171_goldrain:1#text-information`

image

**Comment by gappc** [comment]
DetailView / EditView: opening the page of an additional property directly by inserting the URL into the browser opens the the page correctly, but the menu on the left is not updated correctly

Example link: `BASE_URL//dataset/detail/tourism/v1/ODHActivityPoi/echarging_vek:hyc_150_19bz00171_goldrain:1#echargingdata`

image

**Comment by gappc** [comment]
Strange things happen when the Toolbox in one of the Table/Detail/EditViews is open in desktop width (e.g. 1400px) and the browser window is resized to mobile widths (e.g. 600px):

In that case, nothing is visible anymore, see images below. Note that this does not happen when the Toolbox is closed.

Desktop:
image

Resized window to mobile width:
image

**Comment by gappc** [comment]
Console warnings appear when dataset is opened, seems to be related to Toolbox changes, see image below.

Example URL: `BASE_URL/dataset/table/tourism/v1/Accommodation`

image

**Review by gappc — CHANGES_REQUESTED** [review]
Lots of changes, lots of new stuff, some problems encountered ;) please fix

**Comment by gappc** [comment]
@MatteoBiasi @a-crea thanks for the fixes, should I do another code review?

Another important thing about branch alignment: I saw that the current [development](https://github.com/noi-techpark/it.bz.opendatahub.databrowser) branch and your PR branch [august-2024-ui-and-usability-improvements](https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/tree/feat/august-2024-ui-and-usability-improvements) divert after commit `disabling main deployment` with commit SHA `5517df97392a5d689870470ce581e798d62f3af1`. You can see that when you open [your branch in GitHub](https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/tree/feat/august-2024-ui-and-usability-improvements), where the following is stated (see image below):

"This branch is [30 commits ahead of](https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/compare/noi-techpark%3Ait.bz.opendatahub.databrowser%3Adevelopment...feat/august-2024-ui-and-usability-improvements), [12 commits behind](https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/compare/feat/august-2024-ui-and-usability-improvements...noi-techpark%3Ait.bz.opendatahub.databrowser%3Adevelopment)" 

image

This could be problematic for merging and tooling. I would be very grateful (and strongly recommend) if you could apply your changes on **TOP** of the current `development` branch. 

That way we avoid possible merge problems that go unnoticed. Otherwise there will be double (same) commits (see image below), maybe stuff that was changed and is overwritten by your changes and all kinds of other nasty things. You can take a look at the result if you try to do the merge locally.

image


If you have any problems with rebase / merge or want to discuss this topic, just let me know :+1:

**Comment by gappc** [comment]
@MatteoBiasi @a-crea @xleddyl guys, while i appreciate that you are ironing out this PR, I have some trouble with the way you provide your contributions.

First, this PR becomes very big, touching a lot of stuff. While I myself sometimes go down the unhappy route of big PRs, it remains to say that they are usually problematic in terms of code and feature changes and especially for code review and testing. I thought we agreed to not create massive PRs if there are not very good reasons.

Second, I'd like to ask you for the **THIRD TIME** to bring this PR up do date with the current `development` branch, without polluting the commit history. Or is there any good reason not to do so, that I cannot see?

Third, I see that you introduce / change project dependencies (previously it was a cookie library, now you update tailwind). While updating libraries is fine to some degree, if its not strictly necessary I'd ask you to not do that in such a massive PR. Finding the reason for bugs or degradations becomes hard when everything is mixed together - I know what I'm talking about from different perspectives. If you think you have to change dependencies, lets discuss it briefly.

@sseppi: whats your take on this?

**Comment by MatteoBiasi** [comment]
Hi Christian,
To be honest, last time we created a PR for each issue, and we concluded that it wasn't ideal because there were too many PRs. Now, however, the problem seems to be the opposite. 😅 Actually, I think none of us expected these changes to be so extensive and complex, and we thought that doing everything together in a single PR would be acceptable. But it turned out not to be optimal after all.

Anyway, you were too quick! We hadn’t aligned the PR with development yet because we hadn't finished the changes. We should now have resolved the issue with the diverging history and aligned everything. We couldn’t do it earlier because with the diverging histories, it wasn't possible, so we preferred to focus on fixing the issues first and then on the rebase.

Regarding the library updates, of course, we don’t do them randomly or carelessly. Specifically, we had to update Tailwind due to a glitch that rendered a strange square around the search box. Updating solved the problem. We also checked the changelog before updating, and the update shouldn't break anything existing. Additionally, we thoroughly navigated the platform afterward to check.

Everything should now be good to merge the PR. As for the FIXME comments, we’ve created two specific issues (#600, #601) to allow the rest of the team to complete the work.

**Comment by gappc** [comment]
@MatteoBiasi I appreciate that you and your team are working on the features and, of course, it can happen that a PR gets to big - do you remember my PR from last year? 😉 Nevertheless, experience tells that smaller PRs are better, I think we both agree on that.

Regarding the alignment issues, I noticed that commits continued to accumulate on this PR, even though it had problems with branch alignment. Mentioning that fact twice seemed to get unnoticed.

Now, fixing stuff first, then doing alignment is one possible strategy, but it is harder to get right, simply because you have to review and test all changes again after the alignment.

To get an idea of the current state of the PR, I took a quick look at your new series of commits and it seems that a lot improved.

Unfortunately however, previous, already fixed problems reemerged. For example, your commits introduce the dependency to vue3-cookies in `package.json` anew (see https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/blob/0602feb6ddefed8b58a3fabde8bbc44d12adb783/databrowser/package.json#L37). This was mentioned during the code review and was already fixed by your team. I expect other problems to have reemerged and therefore I'm not as sure as you that this PR is ready to merge. That's my personal opinion, others may see things differently.

In summary I think that we need to discuss some stuff to improve future collaboration, but lets do that in another format, not in this PR here.

**Comment by sseppi** [comment]
@MatteoBiasi @gappc thank you for the contribution of both of you in trying to close this PR. In my point of view there is space for optimization, as always, but now we have to find the best and quickest way to fix and close this issue.

Since is the release of the features included in this PR is quite urgent, what about organizing a pair programming session to review the code together and close the PR?

@gappc is there a way to publish this PR in a dedicated staging environment, to share it with the customer care team and let them test and prepare the documentation we need for the clients?

**Comment by MatteoBiasi** [comment]
@gappc The issues with the build have been resolved. Regarding the cookie package, the problem hadn't actually resurfaced; the code had been corrected using localStorage, but the package was forgotten and not removed. I've now removed it. So, everything is fine on that front as well.

**Comment by gappc** [comment]
@MatteoBiasi thx for the fixes, @sseppi you can find the deployed PR here: https://0.databrowser.gappc.net/

**Comment by sseppi** [comment]
> @MatteoBiasi thx for the fixes, @sseppi you can find the deployed PR here: https://0.databrowser.gappc.net/

@gappc thank you for the link! I shared it with the customer care, to test and document the new featres.

**Comment by sseppi** [comment]
Today while I was testing I found another strange behaviour that seems a bug. I documented it in a dedicated issue:

https://github.com/noi-techpark/it.bz.opendatahub.databrowser/issues/602

@MatteoBiasi can you please have a look at it?

**Comment by MatteoBiasi** [comment]
Thank you @gappc and @sseppi for the two reports. @sseppi, if it's fine with you, shall we give ourselves another 1-2 days to complete the bug reporting from your side and then do a round of fixes?

**Comment by sseppi** [comment]
> Thank you @gappc and @sseppi for the two reports. @sseppi, if it's fine with you, shall we give ourselves another 1-2 days to complete the bug reporting from your side and then do a round of fixes?

@MatteoBiasi from mi side I did some testing yesterday and I inserted everything I found. Being the feature changes that big is difficult to be sure to have tested everything, but we need to go online till the end of the month. For that reason I would suggest to proceed fix everything we found and try to go online.

**Comment by MatteoBiasi** [comment]
@sseppi the fixes for the reported issues have been pushed. For #609, we’ve temporarily hidden the input while waiting for a final decision on what to do. Also, see the note in #602.

**Comment by sseppi** [comment]
@gappc I think the PR is ready to be reviewed

**Comment by gappc** [comment]
@sseppi @MatteoBiasi I've reviewed the changes. Although there are some minor bugs / issues (see below), I believe the PR is ready to be merged..

I would greatly appreciate it if we could address the findings mentioned below, perhaps in a dedicated issue, to ensure they are not overlooked. Lowering the quality of the code will make it harder to maintain and extend the codebase in the future. I'm confident that these issues can be easily fixed, so please don't hesitate to ask me if you need any assistance.

## Bug / Issue list

### Referenced attributes not shown when toggle is disabled

Detail view: referenced attributes are not shown when "Load referenced attributes" is disabled (see feedback from previous code review [here](https://github.com/noi-techpark/it.bz.opendatahub.databrowser/pull/596#issuecomment-2346619459))

### Text for toggle could be misunderstood

Detail view toolbox: the text for the toggle could be misunderstood: "Show attributes which are referenced from other datasets". I think it is the other way around, something like "Show attribute values that are referenced from this record / dataset / (some better naming?)"

### Search button on mobile

On mobile, the search button is still visible in detail / edit / raw view, but clicking on it has no useful effect.  I understand that at the moment the search is not available in these views, but the button should be hidden in that case.

### Popover vs Tooltip component

Popover vs Tooltip component (see feedback from previous code review [here](https://github.com/noi-techpark/it.bz.opendatahub.databrowser/pull/596#discussion_r1756985803)). I think we should try to stick with Popover component for the reasons outlined in the feedback.

### Duplicate attributes shown in Toolbox

Duplicate attributes shown in Toolbox under "Export datasets" in detail view (see image below)

image

### Additional left padding in table view filters

Additional left padding in table view filters (see image below)

image

### Warnings in the console

Warnings in the console when opening the detail view of a record (see image below), e.g. `BASE_URL/dataset/detail/tourism/v1/Accommodation/5E4BEFD440200CE4732A7FA30897C402`

image

### Inconsistent "Show deprecated fields" toggles

In detail / edit view, the "Show deprecated fields" toggles are not in sync (see image below).

image

### Invalid HTML

Self-closing div is not valid HTML (see [DatasetHeaderOverlay.vue](https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/blob/b00ac0eaf39bb5c45b5e8527f8bbde7198e4249a/databrowser/src/domain/datasets/ui/header/DatasetHeaderOverlay.vue#L14))

### TypeScript issues

#### Type casts

Don't do explicit type casts if not strictly necessary.

If a given type is `unknown` and you expect it to be an array, then test for that e.g. `if (Array.isArray(value))` (see problematic code in [SubCategories.vue here](https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/blob/b00ac0eaf39bb5c45b5e8527f8bbde7198e4249a/databrowser/src/domain/datasets/ui/category/SubCategories.vue#L102)). TypeScript will understand what you're trying to do.

Other problematic instances:

- https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/blob/b00ac0eaf39bb5c45b5e8527f8bbde7198e4249a/databrowser/src/domain/datasets/ui/common/load/singleRecordCategories.ts#L64 (explicit type cast to `ComputeSingleRecordCategories`)

- https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/blob/b00ac0eaf39bb5c45b5e8527f8bbde7198e4249a/databrowser/src/domain/datasets/ui/detailView/DetailView.vue#L68 (explicit type cast to `ReferenceInfoToolBoxFetchUrlInfo[]`)

- https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/blob/b00ac0eaf39bb5c45b5e8527f8bbde7198e4249a/databrowser/src/domain/datasets/ui/header/DatasetHeader.vue#L148 (explicit type cast to `any`)

- https://github.com/codeworks-projects/it.bz.opendatahub.databrowser/blob/b00ac0eaf39bb5c45b5e8527f8bbde7198e4249a/databrowser/src/domain/datasets/ui/header/DatasetHeader.vue#L309 (explicit type cast to `string` and `string[]`)

#### Usage of `any`

I saw that new code was introduced that uses the `any` type, e.g. in the file `DatasetHeader.vue`. Please avoid using `any` as much as possible. If you're not sure about the type of a variable, use `unknown` instead of `any`. If you're sure about the type, then use that type. If you have further trouble with TypeScript, then you can ask me any time, I'm confident we'll get the code working without `any` 99,5% of the time.

**Comment by gappc** [comment]
@sseppi @mrabans @MatteoBiasi @a-crea @xleddyl @pkritzinger @Mazzolintis you can view the state of the current PR here: https://0.databrowser.gappc.net/

**Comment by RudiThoeni** [comment]
@sseppi should this PR be merged now or let us wait for this minor issues  (@gappc ) listed to be fixed?
I also fear that they are overlooked if not fixed yet........

**Comment by sseppi** [comment]
@MatteoBiasi how long will it take to fix the issuee mentioned in the yesterday's comment: https://github.com/noi-techpark/it.bz.opendatahub.databrowser/pull/596#issuecomment-2381437866

@gappc it seems that there is some SSL problem, the databrowser the link you shared doesn't load the datasets:
Screenshot from 2024-09-30 17-41-26

**Comment by MatteoBiasi** [comment]
@sseppi we estimate to fix them by Thursday

**Comment by gappc** [comment]
@sseppi you wrote:

> it seems that there is some SSL problem, the databrowser the link you shared doesn't load the datasets:

The error response comes from the ODH testing environment, e.g. [https://api.tourism.testingmachine.eu/v1/MetaData?pagesize=1000](https://api.tourism.testingmachine.eu/v1/MetaData?pagesize=1000) returns with a TLS error.

 The current databrowser test environment seems also not to work, see [https://databrowser.opendatahub.testingmachine.eu/dataset-overview](https://databrowser.opendatahub.testingmachine.eu/dataset-overview). @RudiThoeni any ideas what the problem could be?

@all as I mentioned, in my opinion the PR can be merged in its current state, there are no blockers or major bugs remaining. I would just appreciate if the list of remaining minor bugs / issues is not lost, so we don't forget about them. 

Thanks all for your work and time :+1:

**Review by gappc — APPROVED** [review]

**Comment by RudiThoeni** [comment]
@sseppi @gappc all works again it was a misconfiguration of the test proxy server thx to @clezag for fixing it

**Pull request merged**
